package dev.lokspel.practice.command.practice.arguments;

import dev.lokspel.practice.manager.arena.util.ArenaWorldUtil;
import dev.lokspel.practice.manager.backend.LanguageManager;
import dev.lokspel.practice.manager.inventory.InventoryManager;
import dev.lokspel.practice.manager.server.ServerManager;
import dev.lokspel.practice.util.Common;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.util.StringUtil;

import java.util.ArrayList;
import java.util.List;

public final class LobbyArg {

    private LobbyArg() {}

    public static void run(Player player, String label, String[] args) {
        if (!player.hasPermission("zpp.practice.lobby")) {
            Common.sendMMMessage(player, LanguageManager.getString("COMMAND.PRACTICE.NO-PERMISSION"));
            return;
        }

        if (args.length == 1) {
            if (ServerManager.getLobby() != null) {
                Location lobbyLocation = ServerManager.getLobby();
                player.teleport(lobbyLocation);
            } else {
                Common.sendMMMessage(player, LanguageManager.getString("COMMAND.PRACTICE.ARGUMENTS.LOBBY.LOBBY-NOT-SET").replace("%label%", label));
            }
        } else if (args.length == 2 && args[1].equalsIgnoreCase("set")) {
            Location lobbyLocation = player.getLocation();

            if (!lobbyLocation.getWorld().equals(ArenaWorldUtil.getArenasWorld()) && !lobbyLocation.getWorld().equals(ArenaWorldUtil.getArenasCopyWorld())) {
                ServerManager.getInstance().setLobby(player, lobbyLocation);

                Common.sendMMMessage(player, LanguageManager.getString("COMMAND.PRACTICE.ARGUMENTS.LOBBY.LOBBY-SET"));
            } else {
                Common.sendMMMessage(player, LanguageManager.getString("COMMAND.PRACTICE.ARGUMENTS.LOBBY.NO-ARENAS-WORLD"));
            }
        } else if (args.length == 2 && args[1].equalsIgnoreCase("load")) {
            InventoryManager.getInstance().setLobbyInventory(player, ServerManager.getLobby() != null);
        } else {
            for (String line : LanguageManager.getList("COMMAND.PRACTICE.ARGUMENTS.LOBBY.COMMAND-HELP"))
                Common.sendMMMessage(player, line.replace("%label%", label));
        }
    }

    public static List<String> tabComplete(Player player, String[] args) {
        List<String> arguments = new ArrayList<>();

        if (!player.hasPermission("zpp.practice.lobby")) return arguments;

        if (args.length == 2) {
            arguments.add("set");
            arguments.add("load");

            return StringUtil.copyPartialMatches(args[1], arguments, new ArrayList<>());
        }

        return arguments;
    }

}
